firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -159,66 +159,86 @@ public static FirebaseInstallations getInstance(@NonNull FirebaseApp app) {
-    PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
-    if (persistedFidMissingOrInErrorState(persistedFidEntry)) {
-      throw new FirebaseInstallationsException(
-          "Failed to get existing fid.", FirebaseInstallationsException.Status.CLIENT_ERROR);
+  private Task<String> addGetIdListener() {
+    TaskCompletionSource<String> taskCompletionSource = new TaskCompletionSource<>();
+    StateListener l = new GetIdListener(taskCompletionSource);
+    synchronized (lock) {
+      listeners.add(l);
     }
-    return persistedFidEntry;
+    return taskCompletionSource.getTask();
   }
 
-  private static boolean persistedFidMissingOrInErrorState(PersistedFidEntry persistedFidEntry) {
-    return persistedFidEntry == null
-        || persistedFidEntry.getRegistrationStatus() == RegistrationStatus.REGISTER_ERROR;
+  private Task<InstallationTokenResult> addGetAuthTokenListener() {
+    TaskCompletionSource<InstallationTokenResult> taskCompletionSource =
+        new TaskCompletionSource<>();
+    StateListener l = new GetAuthTokenListener(utils, taskCompletionSource);
+    synchronized (lock) {
+      listeners.add(l);
+    }
+    return taskCompletionSource.getTask();
   }
 
-  @NonNull
-  private static <F, T> Continuation<F, T> orElse(@NonNull Supplier<T> supplier) {
-    return t -> {
-      if (t.isSuccessful()) {
-        return (T) t.getResult();
+  private void triggerOnStateReached(PersistedFidEntry persistedFidEntry) {
+    synchronized (lock) {
+      Iterator<StateListener> it = listeners.iterator();
+      while (it.hasNext()) {
+        StateListener l = it.next();
+        boolean doneListening = l.onStateReached(persistedFidEntry);
+        if (doneListening) {
+          it.remove();
+          break;
+        }
       }
-      return supplier.get();
-    };
+    }
   }
 
-  @NonNull
-  private static <F, T> Continuation<F, T> awaitFidRegistration(
-      @NonNull Supplier<T> supplier, AwaitListener listener) {
-    return t -> {
-      // Waiting for Task that registers FID on the FIS Servers
-      listener.await(AWAIT_TIMEOUT_IN_SECS, TimeUnit.SECONDS);
-      return supplier.get();
+  private final Runnable doRegistration(int authTokenOption) {
+    return () -> {
+      try {
+        PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+
+        // New FID needs to be created
+        if (persistedFidEntry.isErrored() || persistedFidEntry.isNotGenerated()) {
+          String fid = utils.createRandomFid();
+          persistFid(fid);
+          persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+        }
+
+        // GetIdListener will be notified as authTokenOption is always DO_NOT_FORCE_REFRESH.
+        // GetAuthTokenListener should only be notified if FORCE_REFRESH is not required.
+        if (authTokenOption != FORCE_REFRESH) {
+          triggerOnStateReached(persistedFidEntry);
+        }
```

Comment:
is this conditional necessary? couldn't we rely on the final triggerOnStateReached to kick this off?

Reply:
This check is needed to meet the purpose of FORCE_REFRESH. 

If clients request getAuthToken(FORCE_REFRESH) where persisted fid is Reistered and has valid AuthToken, we are still expected to generate a new Auth Token( forcefully refresh it).

Without the check 
``` if (authTokenOption != FORCE_REFRESH) { triggerOnStateReached(persistedFidEntry);} ```, we will notify the client with older auth token, remove the listener from the list and then refresh the auth token and persist it.